profiling/src/profile/internal/string_table.rs

```diff
@@ -0,0 +1,227 @@
+
+    #[allow(unused)]
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Inserts the string into the table, if it did not already exist. The id
+    /// of the string is returned.
+    ///
+    /// # Panics
+    /// Panics if a new string needs to be inserted but the offset of the new
+    /// string doesn't fit into a [StringId].
+    #[inline]
+    pub fn insert(&mut self, str: &str) -> StringId {
+        self.insert_full(str).0
+    }
+
+    /// Inserts the string into the table, if it did not already exist. The id
+    /// of the string is returned, along with whether the string was inserted.
+    ///
+    /// # Panics
+    /// Panics if a new string needs to be inserted but the offset of the new
+    /// string doesn't fit into a [StringId].
+    #[inline]
+    pub fn insert_full(&mut self, str: &str) -> (StringId, bool) {
+        // For performance, delay converting the &str to a String until after
+        // it has been determined to not exist in the set. This avoids
+        // temporary allocations.
+        self.inner
+            .with_dependent_mut(|table| match table.map.get(str) {
+                None => {
+                    let id = StringId::from_offset(table.vec.len());
+                    let bumped_str = table.arena.alloc_str(str);
+
+                    table.vec.push(bumped_str);
+                    table.map.insert(bumped_str, id);
+                    assert_eq!(table.vec.len(), table.map.len());
+                    (id, true)
+                }
+                Some(id) => (*id, false),
+            })
+    }
+
+    /// Gets the string associated with the id.
+    ///
+    /// # Panics
+    /// Panics if the [StringId] doesn't exist in the table.
+    #[inline]
+    pub fn get_id(&self, id: StringId) -> &str {
+        self.inner.with_dependent(|table| {
+            let offset = id.to_offset();
+            match table.vec.get(offset) {
+                Some(str) => str,
+                None => panic!("expected string id {offset} to exist in the string table"),
+            }
+        })
+    }
+
+    #[cfg(test)]
+    #[allow(unused)]
+    #[inline]
+    pub fn get_range(&self, range: Range<usize>) -> &[&str] {
+        self.inner.with_dependent(|table| &table.vec[range])
+    }
+
+    /// Returns an iterator over the strings in the table. The items are
+    /// returned in the order they were inserted, matching the [StringId]s.
+    pub fn iter(&self) -> impl Iterator<Item = &str> {
+        self.inner.with_dependent(|table| table.vec.iter().copied())
+    }
+}
+
+impl Default for StringTable {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /// If this fails, bumpalo may have changed its allocation patterns, and
+    /// [StringTable::new] may need adjusted.
+    #[test]
+    fn test_bump() {
+        let arena = Bump::with_capacity(4000);
+        assert_eq!(4096 - 64, arena.chunk_capacity());
```

Comment:
Where does the 64 come from? Is this guaranteed not to change when Bumpalo is upgraded?

Reply:
There's no guarantee, which is why there's a test -- if we upgrade bumpalo and it's not what we expect, we'll know.

As for the magic numbers, they come from internal details. I guess I could link this:
https://docs.rs/bumpalo/latest/src/bumpalo/lib.rs.html#435-444

```rs
// Maximum typical overhead per allocation imposed by allocators.
const MALLOC_OVERHEAD: usize = 16;

// This is the overhead from malloc, footer and alignment. For instance, if
// we want to request a chunk of memory that has at least X bytes usable for
// allocations (where X is aligned to CHUNK_ALIGN), then we expect that the
// after adding a footer, malloc overhead and alignment, the chunk of memory
// the allocator actually sets aside for us is X+OVERHEAD rounded up to the
// nearest suitable size boundary.
const OVERHEAD: usize = (MALLOC_OVERHEAD + FOOTER_SIZE + (CHUNK_ALIGN - 1)) & !(CHUNK_ALIGN - 1);
```